src/stores/SessionStore.js

```diff
@@ -0,0 +1,61 @@
+import dis from '../dispatcher';
+import EventEmitter from 'events';
+
+/**
+ * A class for storing application state to do with the session. This is a simple flux
+ * store that listens for actions and updates its state accordingly, informing any
+ * listeners (views) of state changes via the 'update' event.
+ */
+class SessionStore extends EventEmitter {
+    constructor() {
+        super();
+
+        // Initialise state
+        this._state = {
+            cachedPassword: localStorage.getItem('mx_pass'),
+        };
+
+        dis.register(this._onAction.bind(this));
+    }
+
+    _update() {
+        // Persist state to localStorage
+        if (this._state.cachedPassword) {
+            localStorage.setItem('mx_pass', this._state.cachedPassword);
+        } else {
+            localStorage.removeItem('mx_pass', this._state.cachedPassword);
+        }
+
+        this.emit('update');
+    }
+
+    _setState(newState) {
+        this._state = Object.assign(this._state, newState);
+        this._update();
+    }
+
+    _onAction(payload) {
+        switch (payload.action) {
+            case 'cached_password':
+                this._setState({
+                    cachedPassword: payload.cachedPassword,
+                });
+                break;
+            case 'password_changed':
+                this._setState({
+                    cachedPassword: null,
+                });
+                break;
+        }
+    }
+
+    getCachedPassword() {
+        return this._state.cachedPassword;
+    }
+}
+
+let singletonSessionStore = null;
+if (!singletonSessionStore) {
+    singletonSessionStore = new SessionStore();
```

Comment:
If it wasn't a singleton we'd have to pass the store down the tree, running into one of the problems this store aims to solve: passing props through components that shouldn't care/passing too many props. I found a nice reddit comment that summarises this in the form of a rebuttal against `... props`: https://www.reddit.com/r/reactjs/comments/4v3mcb/passing_down_too_many_props_to_child_components/d5v8yb7/